src/vs/workbench/contrib/timeline/common/timelineService.ts

```diff
@@ -0,0 +1,207 @@
+		// 	sourceDescription: 'Local History',
+		// 	async provideTimeline(uri: URI, token: CancellationToken) {
+		// 		return [
+		// 			{
+		// 				id: '1',
+		// 				label: 'Undo Timeline1',
+		// 				description: uri.toString(true),
+		// 				date: Date.now()
+		// 			},
+		// 			{
+		// 				id: '2',
+		// 				label: 'Undo Timeline2',
+		// 				description: uri.toString(true),
+		// 				date: Date.now() - 100
+		// 			}
+		// 		];
+		// 	},
+		// 	dispose() { }
+		// });
+	}
+
+	async getTimeline(uri: URI, token: CancellationToken, sources?: Set<string>) {
+		this.logService.trace(`TimelineService#getTimeline(${uri.toString(true)})`);
+
+		const requests = new Map<string, Promise<TimelineItem[] | CancellationErrorWithId<string>>>();
+
+		for (const provider of this._providers.values()) {
+			if (sources && !sources.has(provider.source)) {
+				continue;
+			}
+
+			requests.set(provider.source, provider.provideTimeline(uri, token));
+		}
+
+		// TODO[ECA]: What should the timeout be for waiting for individual providers?
+		const timelines = await raceAll(requests /*, 5000*/);
+
+		const timeline = [];
+		for (const [source, items] of timelines) {
+			if (items instanceof CancellationError) {
+				this.logService.trace(`TimelineService#getTimeline(${uri.toString(true)}) source=${source} cancelled`);
+				continue;
+			}
+
+			if (items.length === 0) {
+				continue;
+			}
+
+			timeline.push(...items.map(item => ({ ...item, source: source })));
+		}
+
+		timeline.sort((a, b) => b.timestamp - a.timestamp);
+		return timeline;
+	}
+
+	registerTimelineProvider(provider: TimelineProvider): IDisposable {
+		this.logService.trace(`TimelineService#registerTimelineProvider: source=${provider.source}`);
+
+		const source = provider.source;
+
+		const existing = this._providers.get(source);
+		if (existing && !existing.replaceable) {
+			throw new Error(`Timeline Provider ${source} already exists.`);
+		}
+
+		this._providers.set(source, provider);
+		if (provider.onDidChange) {
+			this._providerSubscriptions.set(source, provider.onDidChange(uri => this.onProviderTimelineChanged(provider.source, uri)));
+		}
+		this._onDidChangeProviders.fire();
+
+		return {
+			dispose: () => {
+				this._providers.delete(source);
+				this._onDidChangeProviders.fire();
+			}
+		};
+	}
+
+	unregisterTimelineProvider(source: string): void {
+		this.logService.trace(`TimelineService#unregisterTimelineProvider: source=${source}`);
+
+		if (!this._providers.has(source)) {
+			return;
+		}
+
+		this._providers.delete(source);
+		this._providerSubscriptions.delete(source);
+		this._onDidChangeProviders.fire();
+	}
+
+	private onProviderTimelineChanged(source: string, uri: URI | undefined) {
+		// console.log(`TimelineService.onProviderTimelineChanged: source=${source} uri=${uri?.toString(true)}`);
+
+		this._onDidChangeTimeline.fire(uri);
+	}
+}
+
+function* map<T, TMapped>(source: Iterable<T> | IterableIterator<T>, mapper: (item: T) => TMapped): Iterable<TMapped> {
```

Comment:
Even tho this is workbench-only we don't use "modern" JS because the editor still supports IE11. This should be flagged as TS compile error, no clue why not 🤷‍♂

Reply:
This also mean to `Iterable` and such, basically no globals that IE11 doesn't support (and that cannot be poly-filled)